lib/galaxy/datatypes/sniff.py

```diff
@@ -103,28 +103,89 @@ def stream_to_file(stream, suffix='', prefix='', dir=None, text=False, **kwd):
-    >>> open(fname).read()
-    '1 2\\n3 4\\n'
+    >>> def assert_converts_to_1234(content, block_size=1024):
+    ...       fname = get_test_fname('temp.txt')
+    ...       with open(fname, 'w') as fh:
+    ...           _ = fh.write(content)
+    ...       rval = convert_newlines(fname, tmp_prefix="gxtest", tmp_dir=tempfile.gettempdir(), block_size=block_size)
+    ...       assert rval == (2, None), rval
+    ...       actual_contents = open(fname).read()
+    ...       assert '1 2\\n3 4\\n' == actual_contents, actual_contents
+    >>> # Verify ends with newline - with or without that on inputs - for any of
+    >>> # \\r \\n or \\r\\n newlines.
+    >>> assert_converts_to_1234("1 2\\r3 4")
+    >>> assert_converts_to_1234("1 2\\n3 4")
+    >>> assert_converts_to_1234("1 2\\r\\n3 4")
+    >>> assert_converts_to_1234("1 2\\r3 4\\r")
+    >>> assert_converts_to_1234("1 2\\n3 4\\n")
+    >>> assert_converts_to_1234("1 2\\r\\n3 4\\r\\n")
+    >>> assert_converts_to_1234("1 2\\r3 4", block_size=2)
+    >>> assert_converts_to_1234("1 2\\n3 4", block_size=2)
+    >>> assert_converts_to_1234("1 2\\r\\n3 4", block_size=2)
+    >>> assert_converts_to_1234("1 2\\r3 4\\r", block_size=2)
+    >>> assert_converts_to_1234("1 2\\n3 4\\n", block_size=2)
+    >>> assert_converts_to_1234("1 2\\r\\n3 4\\r\\n", block_size=2)
+    >>> assert_converts_to_1234("1 2\\r3 4", block_size=3)
+    >>> assert_converts_to_1234("1 2\\n3 4", block_size=3)
+    >>> assert_converts_to_1234("1 2\\r\\n3 4", block_size=3)
+    >>> assert_converts_to_1234("1 2\\r3 4\\r", block_size=3)
+    >>> assert_converts_to_1234("1 2\\n3 4\\n", block_size=3)
+    >>> assert_converts_to_1234("1 2\\r\\n3 4\\r\\n", block_size=3)
     """
     fd, temp_name = tempfile.mkstemp(prefix=tmp_prefix, dir=tmp_dir)
-    with io.open(fd, mode="wt", encoding='utf-8') as fp:
-        i = None
-        for i, line in enumerate(io.open(fname, encoding='utf-8')):
-            fp.write("%s\n" % line.rstrip("\r\n"))
-    if i is None:
+    i = 0
+    try:
+        with io.open(fd, mode="wt", encoding='utf-8') as fp:
+            with io.open(fname, encoding='utf-8') as fi:
+                partial_line = False
+                while True:
+                    line = fi.readline(block_size)
+                    if not line:
+                        if partial_line:
+                            fp.write(u"\n")
+                            i += 1
+                        break
+
+                    if line[-1] == u"\n":
+                        partial_line = False
+                        fp.write(line)
+                        i += 1
+                        continue
+
+                    # We have a block... maybe at the end of the file.
+                    partial_line = True
+                    fp.write(line)
+    except UnicodeDecodeError:
+        fd, temp_name = tempfile.mkstemp(prefix=tmp_prefix, dir=tmp_dir)
         i = 0
-    else:
-        i += 1
+        with io.open(fd, mode="wb") as fp:
```

Comment:
Is there a case where only doing this would lead wrong results ?

Reply:
https://github.com/techtonik/dos2unix/blob/master/dos2unix.py#L18

Hmm... I don't have confidence to say but I doubt it. This dos2unix Python function does not worry about encoding.